### ISSUE TITLE
UDPSession no longer flushes output queue on Close

### ISSUE THREAD

**Issue by cohosh** [xtaci/kcp-go#273]
Ever since https://github.com/xtaci/kcp-go/commit/651dc4a6a4ff3226775d4df48632498f2125d31f, a `UDPSession` will not flush queued outgoing packets when the connection is closed. Instead, there is a race condition in the `select` call of [`postProcess`](https://github.com/xtaci/kcp-go/blob/0b3010f130ca3f7368c510917b90b215105a5086/sess.go#L610) when the `s.die` channel is closed in `Close`:

```golang
		select {
		case buf := <-s.chPostProcessing: // dequeue from post processing

[snip]
			// notify chCork only when chPostProcessing is empty
			if len(s.chPostProcessing) == 0 {
				select {
				case chCork <- struct{}{}:
				default:
				}
			}
		case <-chCork: // emulate a corked socket
			if len(txqueue) > 0 {
				s.tx(txqueue)
				// recycle
				for k := range txqueue {
					xmitBuf.Put(txqueue[k].Buffers[0])
					txqueue[k].Buffers = nil
				}
				txqueue = txqueue[:0]
			}

		case <-s.die:
			return
		}
```

if `Close` is called immediately after a `Write`, the queued packets will not be sent. This wasn't an issue before because the [`uncork`](https://github.com/xtaci/kcp-go/commit/651dc4a6a4ff3226775d4df48632498f2125d31f#diff-023e5d53117ee139b039639b94ea264e149d748475b3602ba3b676904ca04443L372) function was [called from within `Close`](https://github.com/xtaci/kcp-go/commit/651dc4a6a4ff3226775d4df48632498f2125d31f#diff-023e5d53117ee139b039639b94ea264e149d748475b3602ba3b676904ca04443L398).

I noticed this because [one of our unit tests](https://gitlab.torproject.org/tpo/anti-censorship/pluggable-transports/snowflake/-/blob/97e21e3a29f8dd8306ed893a8341ce91846b02f7/common/turbotunnel/queuepacketconn_test.go#L157) started failing when we [tried to update the version](https://gitlab.torproject.org/tpo/anti-censorship/pluggable-transports/snowflake/-/merge_requests/353) of this library.

**Comment by xtaci** [comment]
https://github.com/xtaci/kcp-go/compare/master...hotfix273 
HI @cohosh, can you try this commit?

**Comment by xtaci** [comment]
v5.6.16

**Comment by cohosh** [comment]
Thanks, that got rid of some of the race conditions but it looks like there are some remaining and our tests are still failing.

There's one [in the output callback](https://github.com/xtaci/kcp-go/blob/1fa6fb277f6fd39d30facbcffca3d7c79023b647/sess.go#L222-L227):
```golang
// delivery to post processing
select {
case sess.chPostProcessing <- bts:
case <-sess.die:
return
}
```
We could probably just remove the select statement here and always send the bytes to `sess.chPostProcessing`.

I just tried this locally but even with this fix there's still another problem where by the time the output callback sends to `sess.chPostProcessing`, the `select` statement in `postProcess` has already died because at the time, there wasn't anything in that channel.

**Comment by xtaci** [comment]
> Thanks, that got rid of some of the race conditions but it looks like there are some remaining and our tests are still failing.
> 
> There's one [in the output callback](https://github.com/xtaci/kcp-go/blob/1fa6fb277f6fd39d30facbcffca3d7c79023b647/sess.go#L222-L227):
> 
> ```go
> // delivery to post processing
> select {
> case sess.chPostProcessing <- bts:
> case <-sess.die:
> return
> }
> ```
> 
> We could probably just remove the select statement here and always send the bytes to `sess.chPostProcessing`.
No,we can't do that, if chPostProcessing is full(because postProcess exits). It blocks forever on waiting for sending.

> 
> I just tried this locally but even with this fix there's still another problem where by the time the output callback sends to `sess.chPostProcessing`, the `select` statement in `postProcess` has already died because at the time, there wasn't anything in that channel.

Yup, The thing is a session cannot wait forever. So If 'Close()' is called, the session makes the 'best-effort delivery'.
